### DISCUSSION TITLE
Compiling LoRaMac-node on Arch Linux

### DISCUSSION THREAD

**Discussion by DavidKrepzi** [Lora-net/LoRaMac-node#1101] [Q&A]
Thank you for enabling gnu utilities so this project can be compiled without any specific IDE but  just using the terminal. This is what I am trying to achieve.  I am following the advice given on https://github.com/Lora-net/LoRaMac-node/blob/master/doc/development-environment.md, step by step.

The hardware I want to compile this LoRaWAN library for is:

1. NUCLEO-L476RG, STM32 Nucleo-64 Development Board
2. https://www.dragino.com/products/lora/item/102-lora-shield.html [SX1276 Radio]

My goal is to get a first basic version to be able to transmit a LoRa frame using this library. I first do a git clone of the project

`git clone https://github.com/Lora-net/LoRaMac-node.git
`

I then follow the instructions to setup the utilities. My exact commands are
```
sudo pacman -S cmake
pacman -S arm-none-eabi-gcc arm-none-eabi-newlib
pacman -S openocd
```

Then I visit the project dir, make a build folder and cd into it.
```
cd LoRaMac-node/
mkdir build
cd build
```

These commands finish successfully. However I receive an error saying the following
```
[user@userHomePC src]$ cmake -DCMAKE_TOOLCHAIN_FILE="cmake/toolchain-arm-none-eabi.cmake" ..
-- The C compiler identification is GNU 10.3.0
-- The CXX compiler identification is GNU 10.3.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/arm-none-eabi-gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/arm-none-eabi-g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
Linker script: /home/user/test/LoRaMac-node/src/boards/NucleoL073/cmsis/arm-gcc/stm32l073xx_flash.ld
-- The ASM compiler identification is GNU
-- Found assembler: /usr/bin/arm-none-eabi-gcc
CMake Error at src/apps/LoRaMac/CMakeLists.txt:46 (message):
  Please turn on Class B support of LoRaMac ( CLASSB_ENABLED=ON ) to use
  periodic-uplink-lpp, fuota-test-01 sub projects

-- Configuring incomplete, errors occurred!
See also "/home/user/test/LoRaMac-node/src/CMakeFiles/CMakeOutput.log".
See also "/home/user/test/LoRaMac-node/src/CMakeFiles/CMakeError.log".

```

I'm curious to know some pointers where I can get started. Thank you.

**Comment by mluis1** [comment]
The [development-environment.md](https://github.com/Lora-net/LoRaMac-node/blob/master/doc/development-environment.md) just provides basic instructions on how to launch `cmake` command. Unfortunately at the moment this document does not provide a fully functional `cmake` command example.

The problem that you are encountering is due to some missing parameters on your `cmake` command.
The listed error means that the `CLASSB_ENABLED` option is not enabled.

Please refer to the README.md file [Building Process](https://github.com/Lora-net/LoRaMac-node#building-process) chapter for an example on how to execute the `cmake` command.
Please find bellow an example to build the `periodic-uplink-lpp` project for `EU868` region (I don't know where you are located) and `NucleoL476` Nucleo board plus `SX1276MB1MAS` mbed shield.

```bash
$ mkdir build
$ cd build
$ cmake -DCMAKE_BUILD_TYPE=Release \
        -DTOOLCHAIN_PREFIX="<replace by toolchain path>" \
        -DCMAKE_TOOLCHAIN_FILE="../cmake/toolchain-arm-none-eabi.cmake" \
        -DAPPLICATION="LoRaMac" \
        -DSUB_PROJECT="periodic-uplink-lpp" \
        -DCLASSB_ENABLED="ON" \
        -DACTIVE_REGION="LORAMAC_REGION_EU868" \
        -DREGION_EU868="ON" \
        -DREGION_US915="OFF" \
        -DREGION_CN779="OFF" \
        -DREGION_EU433="OFF" \
        -DREGION_AU915="OFF" \
        -DREGION_AS923="OFF" \
        -DREGION_CN470="OFF" \
        -DREGION_KR920="OFF" \
        -DREGION_IN865="OFF" \
        -DREGION_RU864="OFF" \
        -DBOARD="NucleoL476" \
        -DMBED_RADIO_SHIELD="SX1276MB1MAS" \
        -DSECURE_ELEMENT="SOFT_SE" \
        -DSECURE_ELEMENT_PRE_PROVISIONED="ON" \
        -DUSE_RADIO_DEBUG="ON" ..
$ make
```
Please note that as you are using an SX1276 Dragino shield you need to add the support for it. You can base your work on the `SX1276MB1LAS` or `SX1276MB1MAS` mbed shields.

In order to get your shield supported you need to modify/create the following files
- Modify `src/CMakeLists.txt` to include your shield (i.e: SX1276DRAGINO)
- Modify `src/boards/NucleoL476/board-config.h` to define which MCU pins are connected to your shield
- Modify `src/boards/NucleoL476/board.c` to call your shield init/deinit API functions.
- Modify `src/boards/NucleoL476/CMakeLists.txt` to add build instructions for your shield
- Create `src/boards/NucleoL476/sx1276dragino-board.c` based on `sx1276mb1las-board.c` or `sx1276mb1mas-board.c` depending on the one which is closest to your shield.
 
Once you have added the support for your shield the `cmake` command `-DMBED_RADIO_SHIELD="SX1276MB1MAS"` parameter must be changed to something like: `-DMBED_RADIO_SHIELD="SX1276DRAGINO"` 

In the future it would be nice if you could post this kind of questions on the project [Discussions](https://github.com/Lora-net/LoRaMac-node/discussions) tab. It is a better place to engage discussions and then we can agree if it is an issue or not.
If you don't mind I move this issue to the [Discussions](https://github.com/Lora-net/LoRaMac-node/discussions) tab.

**Comment by DavidKrepzi** [comment]
Thank you so much for your pointers! And yes, I will keep in mind to move further discussions to the right tab.
